src/kmeans.jl

```diff
@@ -18,74 +24,106 @@ const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
-                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
-                 tol::Real=_kmeans_default_tol,
-                 display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
-
-    m, n = size(X)
-    m2, k = size(centers)
-    m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
+####
+#### Exported kmeans and kmeans! functions
+####
+
+"""
+    kmeans!(centers, X)
+
+Update the current centers `centers` (of size `p x k` where `p` is the dimension and `k` the
+number of centroids) using the samples contained in `X` (of size `p x n` where `n` is the number
+of samples).
+"""
+function kmeans!(centers::AbstractMatrix{<:AbstractFloat}, # in: current centers (p x k)
+                 X::AbstractMatrix{<:Real};                # in: sample matrix (p x n)
+                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing, # in: sample weights (n)
+                 maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                 tol::Real=_kmeans_default_tol,            # in: tolerance of change at convergence
+                 display::Symbol=_kmeans_default_display,  # in: level of display
+                 distance::SemiMetric=SqEuclidean())       # in: function to compute distances
+    p, n = size(X)
+    p2, k = size(centers)
+
+    p == p2 || throw(DimensionMismatch("Inconsistent array dimensions for `X` and `centers`."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
+    if !(weights === nothing)
+        length(weights) == n || throw(DimensionMismatch("Incorrect length of weights."))
+    end
 
     assignments = Vector{Int}(undef, n)
-    costs = Vector{T}(undef, n)
     counts = Vector{Int}(undef, k)
-    cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights, round(Int, maxiter), tol,
-             display_level(display), distance)
+    # check the types to see wheter the updates: centers[i, cj] += X[i, j] * wj
+    # may occur loss of precision through silent casting
+    update_type = float(weights === nothing ? eltype(X) : promote_type(eltype(X), eltype(weights)))
+    update_type <: eltype(centers) || @warn "The type of the centers update ($update_type) is " *
```

Comment:
I'm not aware of any functions that do this. `promote_type` is supposed to return a reasonable type, and `<:` doesn't mean anything in terms of precision, so better just assume the type is OK.

Another way to choose the appropriate type would be `typeof(one(eltype(X)) * one(eltype(weights))`. Then you're certain `X[i, j] * wj` will return an element of the same type as the array.

Reply:
While I'm happy to use the suggested `typeof(one(eltype(X)) * one(eltype(weights))`, there is still  the question of whether that type is wider than that of centers in `kmeans!`. 

There could be `centers` provided in `Float32` and the result of `typeof(one(eltype(X)) * one(eltype(weights)) ==  Float64`. The resulting update will throw an error: 

```
centers = rand(Float32, 2, 2)
centers[1, 1]  = randn() # error
```

What do you suggest? that this remains uncaught until the actual update which would leave to a hard-to-interpret  stack trace? that we error early ? or that we raise a warning like what  I tried to do here?

Reply:
That's not an error at all:
```julia
julia> centers = rand(Float32, 2, 2)
2×2 Array{Float32,2}:
 0.821542  0.58411 
 0.304231  0.636419

julia> centers[1, 1]  = randn() # error
-1.9371986249569961

julia> centers[1, 1]
-1.9371986f0
```

Reply:
hmm weird you're right, I'm not sure what I saw... anyway even without error I'd say the question still holds?

Reply:
Well either you take the eltype of `centers` and assume that's what the user intended, or you allocate a vector with eltype `typeof(one(eltype(X)) * one(eltype(weights))`. But honestly in practice I'm not sure it will make a difference.

Reply:
Ok I removed the weird type check/warning, I agree that at some point we have to assume the user knows that if they feed Float16 centers things might go badly... (and it's unlikely to happen anyway as you pointed out).